### ISSUE TITLE
SSR memory leak caused by global variables (groupIDRegister, reverseRegister)

### ISSUE THREAD

**Issue by casperle** [styled-components/styled-components#2754]
## Environment
*For some reason the envinfo didnt worked, so at least I provide the following.*

node: v12.9.1
yarn: v1.17.3
npm: v6.10.2
styled-components: v5.0.0-beta.8

## Issue
Our setup:
```jsx
...
const styleSheet = new ServerStyleSheet();
const modules = [];

const app = (
  // tslint:disable-next-line
  <Loadable.Capture report={moduleName => modules.push(moduleName)}>
    <ApolloProvider client={client}>
      <StaticRouter context={routerContext} location={url}>
        <StyleSheetManager sheet={styleSheet.instance}>
          <App />
        </StyleSheetManager>
      </StaticRouter>
    </ApolloProvider>
  </Loadable.Capture>
);

try {
  const render = await renderToStringWithData(app); // from react-apollo
  const styleTags = styleSheet.getStyleTags();
  ...
} catch {...}
...
```

High request load is necessary to reproduce this bug in a short period of time. It is needed to provide 5 renders per 1 second. Creating a CPU profile showed that the `toString` method called inside `ServerStyleSheet._emitSheetCSS` takes longer and longer, while the node process is running.
After start:
![image](https://user-images.githubusercontent.com/8900497/65161376-bc606080-da37-11e9-8a88-6166b37cce5f.png)

Around 15 minutes later:
![image](https://user-images.githubusercontent.com/8900497/65161409-d1d58a80-da37-11e9-9de9-d2d174e3d6c9.png)

It goes up to 6 seconds and the node process dies. The issue is caused by global Maps (groupIDRegister, reverseRegister) which are never reset. File `styled-components/src/sheet/GroupIDAllocator.js`

Because of concurrent requests I am not able to reset them after every request is finished, because I may end up with partial rendered styles.
The solution could be to create new instances of these variables on every request. Unfortunately, I do not know the whole codebase, so for you it could be easier to fix it.

Thanks in advance.

**Comment by kitten** [comment]
See comment in: https://github.com/styled-components/styled-components/issues/2753

**Comment by casperle** [comment]
Hi @kitten,
as you adviced in #2753 we searched our codebase and removed all dynamic styled component generation. Checked twice. After testing it the large object heap issue described in #2753 disappeared.
Unfortunately this one (#2754) is still a thing. Here you have some graphs, maybe you will know:

**DefaultGroupedTag.insertRules()** method:
- app start
![image](https://user-images.githubusercontent.com/8900497/65437854-d8437800-de24-11e9-82c7-0a69232f8fa3.png)
- after ~5 min
![image](https://user-images.githubusercontent.com/8900497/65441119-273fdc00-de2a-11e9-82b8-a77ad73b8454.png)

**StyleSheet.toString()** method, where the **outputSheet()** function is called:
- app start
![image](https://user-images.githubusercontent.com/8900497/65440840-a680e000-de29-11e9-95fe-fe30067435c4.png)
- after ~5 min
![image](https://user-images.githubusercontent.com/8900497/65441196-49395e80-de2a-11e9-88ed-e61d3c394308.png)

Metrics:
![image](https://user-images.githubusercontent.com/8900497/65441257-62420f80-de2a-11e9-89dd-7d364e3adeb0.png)
![image](https://user-images.githubusercontent.com/8900497/65441277-6a9a4a80-de2a-11e9-9004-fb7ad20d93b3.png)
![image](https://user-images.githubusercontent.com/8900497/65441290-72f28580-de2a-11e9-8bdf-17969026f73c.png)
![image](https://user-images.githubusercontent.com/8900497/65441313-7ab22a00-de2a-11e9-8686-b0bffe937539.png)
![image](https://user-images.githubusercontent.com/8900497/65441431-aa613200-de2a-11e9-9455-26b2bb8cf5ce.png)
![image](https://user-images.githubusercontent.com/8900497/65441459-b4833080-de2a-11e9-8758-66a6460ac7da.png)

After ~6 mins with the load of 5 requests/sec the nodejs app died.

So you can see that now everything is OK with the heap size. The problem here is with the external memory - used by V8 - and with the RSS - the memory allocated for the process execution. Maybe it is not caused exactly by what I described but again I see here the usage of **DefaultGroupedTag** methods and the **getGroupForId** and **setGroupForId** functions.

Can you please reopen the issue and discuss what we can do? :)

Thanks in advance.

**Comment by kitten** [comment]
This is very interesting! Glad to hear that the other issue is resolved!

It’s weird to hear that one issue is gone due to the removal of dynamic StyledComponents but the other isn’t. Is there any page I could take a look at or more information on how you do SSR?

If I understand this correctly now you’re still seeing more and more GroupIDs over time. Have you tried altering the styled-components file directly to console.log out some of the registered IDs? That may give you a hint of where the additionally registered IDs are coming from

**Comment by casperle** [comment]
@kitten Sorry for the late response. We did exactly what you suggested and found out that in specific area of our application, which is visible for some users only, we forgot some dynamically generated styled component. It was caused by one of our external library, so it wasnt directly in our codebase, thats why we did not found it in first iteration :/.
Sorry for the incorrect info. Thanks for helping me out.
